contracts/pcv/PCVSwapperUniswap.sol

```diff
@@ -0,0 +1,314 @@
+    function getNextAmountReceivedThreshold() external view returns (uint256) {
+      return _getMinimumAcceptableAmountOut(_getExpectedAmountIn());
+    }
+
+    /// @notice Get the decimal normalizer between tokenSpent and tokenReceived
+    ///     e.g. if tokenSpent has 18 decimals and tokenReceived has 6 decimals,
+    //      this function will return 1e12.
+    /// @return the decimal normalizer number.
+    /// @return a boolean to indicate the normalizer direction (does tokenSpent
+    ///     has more decimals than tokenReceived, or the other way around?).
+    function getDecimalNormalizer() external view returns (uint256, bool) {
+      return _getDecimalNormalizer();
+    }
+
+    // =======================================================================
+    // External functions
+    // =======================================================================
+
+    /// @notice Swap tokenSpent for tokenReceived
+    function swap() external override afterTime whenNotPaused {
+      if (tokenBuyLimit != 0) {
+        require(ERC20(tokenReceived).balanceOf(tokenReceivingAddress) < tokenBuyLimit, "PCVSwapperUniswap: tokenBuyLimit reached.");
+      }
+
+      uint256 amountIn = _getExpectedAmountIn();
+      uint256 amountOut = _getExpectedAmountOut(amountIn);
+      uint minimumAcceptableAmountOut = _getMinimumAcceptableAmountOut(amountIn);
+
+      // Check spot price vs oracle price discounted by max slippage
+      // E.g. for a max slippage of 3%, spot price must be >= 97% oraclePrice
+      require(minimumAcceptableAmountOut <= amountOut, "PCVSwapperUniswap: slippage too high.");
+
+      // Reset timer
+      _initTimed();
+
+      // Perform swap
+      ERC20(tokenSpent).safeTransfer(address(pair), amountIn);
+      (uint256 amount0Out, uint256 amount1Out) =
+          pair.token0() == address(tokenSpent)
+              ? (uint256(0), amountOut)
+              : (amountOut, uint256(0));
+      pair.swap(amount0Out, amount1Out, tokenReceivingAddress, new bytes(0));
+
+      // Emit event
+      emit Swap(
+        msg.sender,
+        tokenSpent,
+        tokenReceived,
+        amountIn,
+        amountOut
+      );
+    }
+
+    // =======================================================================
+    // Internal functions
+    // =======================================================================
+
+    /// @notice see external function getNextAmountSpent()
+    function _getExpectedAmountIn() internal view returns (uint256) {
+      uint256 balance = ERC20(tokenSpent).balanceOf(address(this));
+      require(balance != 0, "PCVSwapperUniswap: no tokenSpent left.");
+      return Math.min(maxSpentPerSwap, balance);
+    }
+
+    /// @notice see external function getNextAmountReceived()
+    function _getExpectedAmountOut(uint256 amountIn) internal view returns (uint256) {
+      // Get pair reserves
+      (uint256 _token0, uint256 _token1) = getReserves();
```

Comment:
the `getReserves()` function assumes that one of the two tokens is FEI. I'd rather do pair.getReserves()